File: src/bot/commands/dob.ts
import { CommandInteraction, PermissionFlagsBits, SlashCommandBuilder, time } from '@npm/discord.js';

import { BouncerSlashCommandBuilder, Command } from './@index.ts';

export default class DateOfBirth implements Command {
  public command(): BouncerSlashCommandBuilder {
    return new SlashCommandBuilder()
      .setName('dob')
      .setDescription('Calculate date of birth from given date.')
      .addStringOption((builder) => {
        return builder.setName('date').setDescription('Date in `YYYY-MM-DD` format.').setRequired(true);
      })
      .setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild);
  }

  public async execute(interaction: CommandInteraction) {
    const dateInput = interaction.options.get('date', true).value! as string;

    const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
    if (!dateRegex.test(dateInput)) {
      await interaction.reply({
        content: 'The date should be in `YYYY-MM-DD` format.',
        ephemeral: true,
      });
      return;
    }

    const [year, month, day] = dateInput.split('-').map(Number);

    if (month < 1 || month > 12) {
      await interaction.reply({
        content: 'Invalid month specified. It should be between 1 and 12.',
        ephemeral: true,
      });
      return;
    }

    if (day < 1 || day > daysInMonth(year, month)) {
      await interaction.reply({
        content: `Invalid day specified for month ${month}. It should be between 1 and ${daysInMonth(year, month)}.`,
        ephemeral: true,
      });
      return;
    }

    if (month === 2 && !isLeapYear(year) && day > 28) {
      await interaction.reply({
        content: `Invalid day specified for non-leap year February. It should be between 1 and 28.`,
        ephemeral: true,
      });
      return;
    }

    const dateOfBirth = new Date(year, month - 1, day);

    if (isNaN(dateOfBirth.getTime())) {
      await interaction.reply({
        content: 'Invalid date specified.',
        ephemeral: true,
      });
      return;
    }

    await interaction.reply({
      content: `The date of birth is ${time(dateOfBirth, 'D')} and the age is ${getExactAge(dateOfBirth)}.`,
      ephemeral: true,
    });
  }
}

function isLeapYear(year: number) {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}

function daysInMonth(year: number, month: number) {
  return [31, isLeapYear(year) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31][month - 1];
}

function getExactAge(birthDate: Date): number {
  const currentDate = new Date();
  let age = currentDate.getFullYear() - birthDate.getFullYear();

  if (
    currentDate.getMonth() < birthDate.getMonth() ||
    (currentDate.getMonth() === birthDate.getMonth() && currentDate.getDate() < birthDate.getDate()) ||
    (isLeapYear(currentDate.getFullYear()) && currentDate.getMonth() === 1 && currentDate.getDate() === 29 &&
      birthDate.getMonth() === 1 && birthDate.getDate() === 29)
  ) {
    age--;
  }

  return age;
}

File: src/bot/commands/dob.rs
use chrono::{format::ParseErrorKind, Datelike, NaiveDate, Utc};
use serenity::all::{CommandOptionType, CreateCommand, CreateCommandOption};
use tracing::error;

use super::BouncerCommand;

use crate::bot::{
    extensions::resolved_options::ResolvedOptionExt,
    helpers::interaction_context::CommandInteractionContext, BouncerState,
};

pub struct Command;
impl<'a> BouncerCommand<'a> for Command {
    const COMMAND_NAME: &'a str = "dob";
    const COMMAND_DESCRIPTION: &'a str = "Calculate the age from given date of birth time.";

    fn command() -> CreateCommand<'a> {
        CreateCommand::new(Self::COMMAND_NAME)
            .description(Self::COMMAND_DESCRIPTION)
            .add_option(
                CreateCommandOption::new(
                    CommandOptionType::String,
                    "date",
                    "Date in `YYYY-MM-DD` format.",
                )
                .required(true),
            )
    }

    async fn execute(
        interaction_context: CommandInteractionContext<'_>,
        _state: &BouncerState,
    ) -> anyhow::Result<()> {
        let Some(date_input) = interaction_context.options.get_string_option("date") else {
            interaction_context
                .reply_string("Please enter a date.", Some(true))
                .await?;
            return Ok(());
        };

        let parsed_date = match NaiveDate::parse_from_str(date_input, "%Y-%m-%d") {
            Ok(date) => date,
            Err(error) => {
                let reply_string = match error.kind() {
                    ParseErrorKind::BadFormat => "The date should be entered in `YYYY-MM-DD` format.",
                    ParseErrorKind::OutOfRange => "Ensure the month is between 1 and 12, and the day is valid for the given month.",
                    ParseErrorKind::TooShort => "The date is incomplete. Please fill in the missing fields.",
                    _ => {
                        error!(
                            "an unexpected error occurred while parsing the date: {:?}",
                            error
                        );

                        "An unexpected error occurred while parsing the date."
                    }
                }.to_string();

                interaction_context
                    .reply_string(reply_string, Some(true))
                    .await?;

                return Ok(());
            }
        };

        let today = Utc::now();
        let mut age = today.year() - parsed_date.year();
        if today.month() < parsed_date.month()
            || today.month() == parsed_date.month() && today.day() < parsed_date.day()
        {
            age -= 1;
        }

        interaction_context
            .reply_string(format!("{age}"), Some(true))
            .await?;

        Ok(())
    }
}
